### PR TITLE
Feature/jaeger support

### PR DESCRIPTION
**Pull request by aspectom**
Adding support for jaeger as a storage backend.

### CODE REVIEW

File: packages/malabi/src/remote-runner-integration/fetch-remote-telemetry.ts [hunk 1]
```diff
@@ -11,16 +14,26 @@ interface FetchRemoteTelemetryProps {
...
                 return res;
             },
         });
-        const protoFormatted = collectorTraceV1Transform.fromJsonEncodedProtobufFormat(res.data);
-        const spans = collectorTraceV1Transform.fromProtoExportTraceServiceRequest(protoFormatted);
+
+        let spans;
+
+        if (process.env.MALABI_STORAGE_BACKEND === StorageBackend.Jaeger) {
+            const spansInJaegerFormat = JSON.parse(res.data).filter(({ traceID }) => traceID === currentTestTraceID)[0].spans;
```
Comment (blumamir) [inline packages/malabi/src/remote-runner-integration/fetch-remote-telemetry.ts:30]: filter -> find

File: packages/malabi/src/remote-runner-integration/fetch-remote-telemetry.ts [hunk 2]
```diff
@@ -11,16 +14,26 @@ interface FetchRemoteTelemetryProps {
...
+            spans = spansInJaegerFormat.map(jaegerSpan => convertJaegerSpanToOtelReadableSpan(jaegerSpan));
```
Comment (blumamir) [inline packages/malabi/src/remote-runner-integration/fetch-remote-telemetry.ts:31]: consider always returning otel data from service. then the test runner is agnostic to the storage backend

File: packages/malabi/src/exporter/jaeger.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { JaegerExporter } from '@opentelemetry/exporter-jaeger';
+
+const JAEGER_HOST = process.env.MALABI_JAEGER_HOST || 'localhost';
+export const jaegerExporter = new JaegerExporter({
+    tags: [],
+    host: JAEGER_HOST,
```
Comment (blumamir) [inline packages/malabi/src/exporter/jaeger.ts:6]: prefer to configure using the opentelemetry environment variables:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#jaeger-exporter

File: packages/malabi/src/exporter/jaeger.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+});
+
+export const getJaegerSpans = async (serviceName: string) => {
+    const JAEGER_PROTOCOL = process.env.MALABI_JAEGER_QUERY_PROTOCOL || 'http';
+    const axios = require('axios');
+    const res = await axios.get(`${JAEGER_PROTOCOL}://${JAEGER_HOST}:16686/api/traces?service=${serviceName}`)
```
Comment (blumamir) [inline packages/malabi/src/exporter/jaeger.ts:12]: keep those, as they might be different from the exporter config. document how to use

Comment (blumamir) [inline packages/malabi/src/exporter/jaeger.ts:12]: you can use axios params to encode url. this is considered safer as it url encode the data and escape problematic input

File: packages/instrumentation-node/package.json [hunk 5]
```diff
@@ -29,24 +29,24 @@
...
-        "opentelemetry-instrumentation-amqplib": "0.23.0",
-        "opentelemetry-instrumentation-aws-sdk": "0.23.0",
-        "opentelemetry-instrumentation-elasticsearch": "0.23.0",
-        "opentelemetry-instrumentation-express": "0.23.0",
-        "opentelemetry-instrumentation-kafkajs": "0.23.0",
-        "opentelemetry-instrumentation-mongoose": "0.23.0",
-        "opentelemetry-instrumentation-neo4j": "0.23.0",
-        "opentelemetry-instrumentation-sequelize": "0.23.0",
-        "opentelemetry-instrumentation-typeorm": "0.23.0"
+        "@opentelemetry/instrumentation-aws-sdk": "^0.4.0",
```
Comment (blumamir) [inline packages/instrumentation-node/package.json:34]: remove caret

File: examples/tests-runner/package.json [hunk 6]
```diff
@@ -6,15 +6,20 @@
...
     "scripts": {
-        "test:example": "jest"
+        "test": "ts-mocha --paths \"./test/*.ts\" --require \"./test/tracing.ts\" --timeout 10000",
+        "test:inmemory": "MALABI_STORAGE_BACKEND=InMemory yarn test",
+        "test:jaeger": "MALABI_STORAGE_BACKEND=Jaeger MALABI_JAEGER_HOST=localhost yarn test"
     },
     "dependencies": {
-        "axios": "^0.21.1"
+        "axios": "^0.21.1",
+        "typescript": "^4.2.4"
```
Comment (blumamir) [inline examples/tests-runner/package.json:15]: can be in dev

File: packages/malabi/src/instrumentation/index.ts [hunk 7]
```diff
@@ -46,3 +86,26 @@ export const instrument = () => {
...
+ * @category Main Functions
+ * @param callback an async function containing all of the current test's span generating operations(API calls etc)
+ */
+export const malabi = async (callback): Promise<TelemetryRepository> => {
+    return new Promise((resolve) => {
+        const tracer = trace.getTracer('malabiManualTracer');
+
+        tracer.startActiveSpan('malabiRoot', async (span) => {
+            const currTraceID = span.spanContext().traceId;
+            await callback();
```
Comment (blumamir) [inline packages/malabi/src/instrumentation/index.ts:101]: wrap in try catch and add a test to verify that exceptions inside the callback are propagated correcly outside the wrapper

File: packages/malabi/src/instrumentation/index.ts [hunk 8]
```diff
@@ -46,3 +86,26 @@ export const instrument = () => {
...
+/**
+ * A wrapper that handles creating a span per test run. returns the spans that were created ready for assertion.
+ * @category Main Functions
+ * @param callback an async function containing all of the current test's span generating operations(API calls etc)
+ */
+export const malabi = async (callback): Promise<TelemetryRepository> => {
+    return new Promise((resolve) => {
+        const tracer = trace.getTracer('malabiManualTracer');
+
+        tracer.startActiveSpan('malabiRoot', async (span) => {
```
Comment (blumamir) [inline packages/malabi/src/instrumentation/index.ts:99]: consider adding some data on which test is running

File: packages/malabi/src/instrumentation/index.ts [hunk 9]
```diff
@@ -46,3 +86,26 @@ export const instrument = () => {
...
 };
+
+/**
+ * A wrapper that handles creating a span per test run. returns the spans that were created ready for assertion.
+ * @category Main Functions
+ * @param callback an async function containing all of the current test's span generating operations(API calls etc)
+ */
+export const malabi = async (callback): Promise<TelemetryRepository> => {
+    return new Promise((resolve) => {
+        const tracer = trace.getTracer('malabiManualTracer');
```
Comment (blumamir) [inline packages/malabi/src/instrumentation/index.ts:97]: send package name of malabi as in package.json "name", and set the version if possible

File: packages/malabi/src/instrumentation/index.ts [hunk 10]
```diff
@@ -46,3 +86,26 @@ export const instrument = () => {
...
+
+        tracer.startActiveSpan('malabiRoot', async (span) => {
+            const currTraceID = span.spanContext().traceId;
+            await callback();
+            span.end();
+            const telemetry = await fetchRemoteTelemetry({
+                portOrBaseUrl: 18393,
```
Comment (blumamir) [inline packages/malabi/src/instrumentation/index.ts:104]: consider changing to env variable or function configuration

**Review by blumamir — APPROVED** [review]

File: packages/malabi/src/instrumentation/index.ts (hunk 9 above)
Comment (aspectom) [reply packages/malabi/src/instrumentation/index.ts:97]: Added this as an issue for later on https://github.com/aspecto-io/malabi/issues/27

**Pull request merged**